File: verify/github.go
package verify

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"strings"
)

// GitHubVerifier verifies GitHub webhook signatures using HMAC-SHA256.
type GitHubVerifier struct {
	secret []byte
}

// NewGitHubVerifier creates a new GitHubVerifier with the given shared secret.
func NewGitHubVerifier(secret string) *GitHubVerifier {
	return &GitHubVerifier{secret: []byte(secret)}
}

// Verify checks the X-Hub-Signature-256 header against the request body.
// Returns an error if the header is missing or malformed.
// Returns false (not error) if the signature does not match.
func (v *GitHubVerifier) Verify(body []byte, headers http.Header) (bool, error) {
	sig := headers.Get("X-Hub-Signature-256")
	if sig == "" {
		return false, fmt.Errorf("missing X-Hub-Signature-256 header")
	}

	if !strings.HasPrefix(sig, "sha256=") {
		return false, fmt.Errorf("malformed X-Hub-Signature-256 header: missing sha256= prefix")
	}

	receivedHex := sig[len("sha256="):]
	received, err := hex.DecodeString(receivedHex)
	if err != nil {
		return false, fmt.Errorf("malformed X-Hub-Signature-256 header: invalid hex: %w", err)
	}

	mac := hmac.New(sha256.New, v.secret)
	mac.Write(body)
	expected := mac.Sum(nil)

	return hmac.Equal(expected, received), nil
}

File: src/verify/github.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import type { IncomingHttpHeaders } from "node:http";
import type { SignatureVerifier } from "./verifier.js";

/**
 * GitHubVerifier verifies GitHub webhook signatures using HMAC-SHA256.
 * Reads the X-Hub-Signature-256 header which has format: sha256=<hex>
 */
export class GitHubVerifier implements SignatureVerifier {
  private secret: string;

  constructor(secret: string) {
    this.secret = secret;
  }

  verify(
    body: Buffer,
    headers: IncomingHttpHeaders
  ): { valid: boolean; error?: string } {
    const sig = headers["x-hub-signature-256"] as string | undefined;
    if (!sig) {
      return { valid: false, error: "missing X-Hub-Signature-256 header" };
    }

    if (!sig.startsWith("sha256=")) {
      return {
        valid: false,
        error:
          "malformed X-Hub-Signature-256 header: missing sha256= prefix",
      };
    }

    const receivedHex = sig.slice("sha256=".length);

    // Validate hex string
    let received: Buffer;
    try {
      received = Buffer.from(receivedHex, "hex");
      // Buffer.from doesn't throw on invalid hex, it just ignores bad chars.
      // Verify the round-trip to catch malformed hex.
      if (received.toString("hex") !== receivedHex.toLowerCase()) {
        return {
          valid: false,
          error:
            "malformed X-Hub-Signature-256 header: invalid hex",
        };
      }
    } catch {
      return {
        valid: false,
        error:
          "malformed X-Hub-Signature-256 header: invalid hex",
      };
    }

    const expected = createHmac("sha256", this.secret).update(body).digest();

    // Use timing-safe comparison to prevent timing attacks
    if (expected.length !== received.length) {
      return { valid: false };
    }

    return { valid: timingSafeEqual(expected, received) };
  }
}
